# Case 1 - Able
* Kyle Anderson U0985830

# Table of Contents

1. [Introduction and Understanding of the Case](#Introduction-and-Understanding-of-the-Case)
2. [Data Import](#Data-Import)
3. [Question 1](#Question-1)
4. [Question 2](#Question-2)
5. [Question 3](#Question-3)
6. [Question 4](#Question-4)
7. [Question 5](#Question-5)
8. [Question 6](#Question-6)
9. [Question 7](#Question-7)

## Introduction and Understanding of the Case


### Problem
* Able's vendors want to understand the impact of when a price drops, will it boost sales enough to offset lower margins. The focus is kid's toy sellers during the holiday season. The goal is to help vendors make informed decisions about their sales strategies by understanding the true effect of price changes on customer behavior.

### Testing and Results
* Is there a benefit to selling at a lower cost which is hypothesized to drive more purchases thus higher revenue than the regular price?
* What is the dollar impact of lowering prices on units sold?

### Import Libraries

In [91]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

### Available Features
* store (numeric): the store’s unique identifier (ID).
* weeks_to_xmas (numeric): a count of the weeks to xmas.
* avg_week_sale (numeric): the size of each business in terms of average products sold per week during that year.
#### Target of IS on sale (1) or NOT on sale (0)
* is_on_sale (boolean): indicator for whether the business had a sale during a the week.
#### Volume sold
* weekly_amount_sold (numeric): the average weekly sales of that store during the week.

## Data Import

In [92]:
# Observed Toy Sales
toys_git = "https://raw.githubusercontent.com/jefftwebb/data/main/toy_sales_data.csv"
df_o = pd.read_csv(toys_git, index_col=False)

# Potential Toy Sales
toy_sales = "https://raw.githubusercontent.com/jefftwebb/data/main/toy_sales_PO.csv"
df_p = pd.read_csv(toy_sales, index_col=False)

## Inspect the data for both data frames

In [93]:
print('Observed')
df_o.head(10)

# This is the actual observed data with the different stores and their output of weekly sold

Observed


store  weeks_to_xmas  avg_week_sales  is_on_sale  weekly_amount_sold
0      1              3           12.98           1              212.08
1      1              2           12.98           1              186.12
2      1              1           12.98           1              160.16
3      1              0           12.98           0               84.20
4      2              3           19.92           0               98.85
5      2              2           19.92           0               59.01
6      2              1           19.92           0               19.17
7      2              0           19.92           0                0.00
8      3              3           18.54           0               93.37
9      3              2           18.54           0               56.29

In [94]:
# Insect Potential
print('Potential')
df_p.head(10)

# This is the potential output if we had the opportunity to see the impact of sale or no sale at the same time

Potential


store  weeks_to_xmas  avg_week_sales  is_on_sale      y0      y1  \
0      1              3           12.98           1  162.08  212.08   
1      1              2           12.98           1  136.12  186.12   
2      1              1           12.98           1  110.16  160.16   
3      1              0           12.98           0   84.20  134.20   
4      2              3           19.92           0   98.85  148.85   
5      2              2           19.92           0   59.01  109.01   
6      2              1           19.92           0   19.17   69.17   
7      2              0           19.92           0    0.00   29.33   
8      3              3           18.54           0   93.37  143.37   
9      3              2           18.54           0   56.29  106.29   

   weekly_amount_sold  
0              212.08  
1              186.12  
2              160.16  
3               84.20  
4               98.85  
5               59.01  
6               19.17  
7                0.00  
8               93.37  
9               56.29

## Question 1
* What is the true ATE of having a sale? This is a calculation (not an estimate) that uses the potential outcomes data, toy_sales_PO.csv, and is defined as: 

In [95]:
# Using the Potentials, lets calculate ATE
# Sale - No sale
df_p['ATE'] = (df_p['y1'] - df_p['y0'])

# Get the sum of column ATE and get the average
result = (df_p['ATE'].sum())/len(df_p['ATE'])
print('Potential ATE')
print(result)

# 45+ means the Treatment of sale is positive!

Potential ATE
45.90093


## Answer 1:
* 45 units increase when a Sale is used

## Question 2
* What is the estimated ATE of having a sale using the observed data, toy_sales.csv? The estimate consists in the difference between the two group averages:

In [96]:
# Create new columns for the exclusive 1 and 0
df_o['Sale'] = df_o.loc[df_o['is_on_sale'] == 1, 'weekly_amount_sold']
df_o['No Sale'] = df_o.loc[df_o['is_on_sale'] == 0, 'weekly_amount_sold']
# print(df_o.columns)
observed_effect = ((df_o['Sale'].mean()) - (df_o['No Sale'].mean()))
print('Observed ATE')
print(observed_effect)


Observed ATE
77.94207879738728


## Answer 2:
* 77.9 Units increase when a sale occurs in the observed data

## Question 3
* What is the bias involved in using the observed data to estimate ATE?

## Answer 3
* As discussed in class with the twins topic. There needs to be randomization in the study as the store that runs sales more or less often would have impacts based on their history. There needs to be more data from previous years. Store size, location, unit price and marketing can all impact the results of the observed results. We only have access to Sale/No Sale.

## Question 4
* A balance table is used to check whether the treatment and control groups in a study are exchangeable based on pre-treatment characteristics. Create a balance table using the observed data (toy_sales.csv) that shows how company size varies by treatment status. Comment on the exchangeability of the groups.

In [97]:
df_balance = df_o.groupby('is_on_sale')['avg_week_sales'].agg(['mean', 'std', 'count']).reset_index()
df_balance.columns = ['is_on_sale', 'Avg_Week_Sales', 'Std_Avg_Week_Sales', 'Count']
total_count = df_balance['Count'].sum()
print("Total Sales Occurrences")
print(total_count)
print("% No Sales Occurrences")
no_sale_count = (df_balance.loc[df_balance['is_on_sale'] == 0, 'Count'].values[0])/total_count
print(no_sale_count)
print("% Sales Occurrences")
sale_count = (df_balance.loc[df_balance['is_on_sale'] == 1, 'Count'].values[0])/total_count
print(sale_count)

# (21.58-18.77)/18.77 = 14% more sales than no sales
# Despite, Sales (1) having 3% more in terms of total occurrences, they had 14% more in avg weekly sales

df_balance.head()


Total Sales Occurrences
2000
% No Sales Occurrences
0.488
% Sales Occurrences
0.512


is_on_sale  Avg_Week_Sales  Std_Avg_Week_Sales  Count
0           0       18.777121            3.716006    976
1           1       21.584229            4.991820   1024

## Answer 4
* They are not exchangable based on the avg sales and std. The standard deviation shows the most concern as a std of 4.99 indicates more variation.

## Question 5
* Company size confounds the relationship between weekly_amount_sold and is_on_sale in the observed data (toy_sales.csv). Create a statistical model to adjust for the confounding. Hint: use linear regression. Report an adjusted ATE. How does it compare to the true ATE? Explain how the adjustment is working.

In [98]:
# Get scikit packages
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import StandardScaler

# Get your independent and dependent variables. If sale, what happens
X = df_o[['is_on_sale', 'avg_week_sales']]
y = df_o['weekly_amount_sold']

model = LinearRegression()
model.fit(X, y)

adjusted_ATE = model.coef_[0] # pull the first value from the attribute
print(f"Adjusted ATE of having a sale)")
print(adjusted_ATE)

# Pull scikit attributes for slope and intercept
print(f"Intercept")
print(model.intercept_)
print(f"Coefficients:")
print(model.coef_)


Adjusted ATE of having a sale)
68.98671031000106
Intercept
2.897194701296044
Coefficients:
[68.98671031  3.19024766]


## Answer 5
* 68.99: For each unit increase in is_on_sale (0 or 1), weekly_amount_sold is expected to increase by approximately 68.99 units

* 3.19: For each unit increase in avg_week_sales, weekly_amount_sold is expected to increase by approximately 3.19 units

## Question 6
* The conditional average treatment effect or CATE is defined as the average treatment effect for a subgroup of the data. Think of weeks_to_xmas as defining subgroups with levels equal to 3, 2, 1, 0. What is the true ATE of a sales campaign conditional on weeks to Christmas?

* Use toy_sales_PO.csv for the calculation.
* You should report 4 numbers and make a comment.

In [99]:
# Pull the levels Weeks to xmas from the Potentials dataset
print(df_p['weeks_to_xmas'].unique())

df_p['ATE'] = df_p['y1'] - df_p['y0']
# CATE, for each level. So it is just easier to do .mean()
cate = df_p.groupby('weeks_to_xmas')['ATE'].mean()

actual = []
for week, metric in cate.items():
    print(fr"Weeks to Christmas: {week}, actual ATE: {round(metric,2)}")
    actual.append(round(metric,2))

[3 2 1 0]
Weeks to Christmas: 0, actual ATE: 38.75
Weeks to Christmas: 1, actual ATE: 46.14
Weeks to Christmas: 2, actual ATE: 48.94
Weeks to Christmas: 3, actual ATE: 49.78


## Question 7
* Use a statistical model to estimate the adjusted ATE of a sales campaign conditional on weeks to Christmas. Hint: add an interaction term to your earlier regression model.

* Challenge: create a plot that illustrates the statistical result.

In [100]:
# The interaction term is going to be the weeks out from xmas and the sale/no sale

df_o['interaction'] = df_o['is_on_sale'] * df_o['weeks_to_xmas']

X = df_o[['is_on_sale', 'weeks_to_xmas', 'interaction']]
y = df_o['weekly_amount_sold']
model = LinearRegression()
model.fit(X, y)

adjusted_cate = {}
for week in sorted(df_o['weeks_to_xmas'].unique()): # I got to the end and figured out that I was doing the inverse of question 6 where it was starting from week 3 and not 0.
    # Sale
    X_sale = np.array([[1, week, week]])
    y_sale = model.predict(X_sale)[0]
    # No sale
    X_no_sale = np.array([[0, week, 0]])
    y_no_sale = model.predict(X_no_sale)[0]
    # No mean since it is specifc to each week
    adjusted_cate[week] = y_sale - y_no_sale

# Adjusted cate results and the difference to actual

x=0 # starting at week 0 
for week, cate in adjusted_cate.items():
    print(f"Weeks to Christmas: {week}, Adjusted ATE: {round(cate,2)}, difference from adjusted to actual: {round(cate - actual[x],2)}")
    x += 1

Weeks to Christmas: 0, Adjusted ATE: 52.75, difference from adjusted to actual: 14.0
Weeks to Christmas: 1, Adjusted ATE: 59.69, difference from adjusted to actual: 13.55
Weeks to Christmas: 2, Adjusted ATE: 66.63, difference from adjusted to actual: 17.69
Weeks to Christmas: 3, Adjusted ATE: 73.57, difference from adjusted to actual: 23.79


C:\Users\kyle.anderson\pythonProject\MKTG6600\MKTG6600\.venv\Lib\site-packages\sklearn\base.py:493: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(
C:\Users\kyle.anderson\pythonProject\MKTG6600\MKTG6600\.venv\Lib\site-packages\sklearn\base.py:493: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(
C:\Users\kyle.anderson\pythonProject\MKTG6600\MKTG6600\.venv\Lib\site-packages\sklearn\base.py:493: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(
C:\Users\kyle.anderson\pythonProject\MKTG6600\MKTG6600\.venv\Lib\site-packages\sklearn\base.py:493: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(
C:\Users\kyle.anderson\pythonProject\MKTG6600\MKTG6600\.venv\Lib\site-packages\sklearn\base.py:493: UserWarning: X does not have

## Answer 7
* The Adjusted CATE is higher because we are overestimating the impacts of weeks out from Christmas and sales. My assumption is that the model believes there is a strong causation to weeks out and increase in sales, which is why it is blown out of proportion.
* I think we could use other features to help with the model, because we don't have the specifics of the store and we are getting confounding. There are so many reasons that Christmas would influence purchases and customer behavior. The purchasing based on weeks out on Christmas doesn't make a huge difference because people are already buying. Maybe the toy store is the only one for hundreds of miles and any sale around Christmas would be thrown out of proportion.